File: libs/power-log-parser/src/lib/parsers/mercenaries-queued-ability-parser.ts
import { GameTag } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventProvider, GameEventHelper } from '../game-event';
import { Node, NodeType, TagChange } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class MercenariesQueuedAbilityParser implements ActionParser {
	readonly ParserName = 'MercenariesQueuedAbilityParser';

	private GameState: GameState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, stateFacade: StateFacade) {
		this.GameState = parserState.GameState;
		this.StateFacade = stateFacade;
	}

	AppliesOnNewNode(node: Node, stateType: StateType): boolean {
		if (stateType !== StateType.PowerTaskList || node.Type !== NodeType.TagChange) {
			return false;
		}
		const tagChange = node.Object as TagChange;
		return (
			tagChange.Name === (GameTag.LETTUCE_ABILITY_TILE_VISUAL_ALL_VISIBLE as number) ||
			tagChange.Name === (GameTag.LETTUCE_ABILITY_TILE_VISUAL_SELF_ONLY as number)
		);
	}

	AppliesOnCloseNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	CreateGameEventProviderFromNew(node: Node): GameEventProvider[] | null {
		const tagChange = node.Object as TagChange;
		const entity = this.GameState.CurrentEntities.get(tagChange.Entity)!;
		const cardId = entity.CardId;
		const controllerId = entity.GetEffectiveController();

		if (tagChange.Value === 0) {
			return [
				GameEventProvider.Create(
					tagChange.TimeStamp,
					'MERCENARIES_ABILITY_UNQUEUED',
					GameEventHelper.CreateProvider(
						'MERCENARIES_ABILITY_UNQUEUED',
						cardId,
						controllerId,
						entity.Id,
						this.StateFacade,
						null,
					),
					true,
					node,
				),
			];
		} else {
			const abilityEntityId = tagChange.Value;
			const abilityEntity = this.GameState.CurrentEntities.has(abilityEntityId)
				? this.GameState.CurrentEntities.get(abilityEntityId)!
				: null;
			const abilityCardId = abilityEntity?.CardId;
			const abilitySpeed = abilityEntity?.GetTag(GameTag.COST) ?? 0;
			if (
				controllerId === this.StateFacade.LocalPlayer?.PlayerId &&
				tagChange.Name === (GameTag.LETTUCE_ABILITY_TILE_VISUAL_ALL_VISIBLE as number)
			) {
				return null;
			}
			return [
				GameEventProvider.Create(
					tagChange.TimeStamp,
					'MERCENARIES_ABILITY_QUEUED',
					GameEventHelper.CreateProvider(
						'MERCENARIES_ABILITY_QUEUED',
						cardId,
						controllerId,
						entity.Id,
						this.StateFacade,
						{
							AbillityEntityId: abilityEntityId,
							AbilityCardId: abilityCardId,
							AbilitySpeed: abilitySpeed,
						},
					),
					true,
					node,
				),
			];
		}
	}

	CreateGameEventProviderFromClose(_node: Node): GameEventProvider[] | null {
		return null;
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/MercenariesQueuedAbilityParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;

namespace HearthstoneReplays.Events.Parsers
{
    public class MercenariesQueuedAbilityParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public MercenariesQueuedAbilityParser(ParserState ParserState, StateFacade helper)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = helper;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            TagChange tagChange = null;
            return stateType == StateType.PowerTaskList
                && node.Type == typeof(TagChange)
                && ((tagChange = node.Object as TagChange).Name == (int)GameTag.LETTUCE_ABILITY_TILE_VISUAL_ALL_VISIBLE
                    || tagChange.Name == (int)GameTag.LETTUCE_ABILITY_TILE_VISUAL_SELF_ONLY);
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return false;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            var tagChange = node.Object as TagChange;
            var entity = GameState.CurrentEntities[tagChange.Entity];
            var cardId = entity.CardId;
            var controllerId = entity.GetEffectiveController();
            if (tagChange.Value == 0)
            {
                return new List<GameEventProvider> { GameEventProvider.Create(
                    tagChange.TimeStamp,
                    "MERCENARIES_ABILITY_UNQUEUED",
                    GameEvent.CreateProvider(
                        "MERCENARIES_ABILITY_UNQUEUED",
                        cardId,
                        controllerId,
                        entity.Id,
                        StateFacade,
                        null),
                    true,
                    node) };
            }
            else
            {
                var abilityEntityId = tagChange.Value;
                var abilityEntity = GameState.CurrentEntities.ContainsKey(abilityEntityId) ? GameState.CurrentEntities[abilityEntityId] : null;
                var abilityCardId = abilityEntity?.CardId;
                var abilitySpeed = abilityEntity.GetTag(GameTag.COST);
                // Only queue the player's ability when they queue it themselves, not when they are revealed
                if (controllerId == StateFacade.LocalPlayer?.PlayerId && tagChange.Name == (int)GameTag.LETTUCE_ABILITY_TILE_VISUAL_ALL_VISIBLE)
                {
                    return null;
                }
                return new List<GameEventProvider> { GameEventProvider.Create(
                    tagChange.TimeStamp,
                    "MERCENARIES_ABILITY_QUEUED",
                    GameEvent.CreateProvider(
                        "MERCENARIES_ABILITY_QUEUED",
                        cardId,
                        controllerId,
                        entity.Id,
                        StateFacade,
                        //null,
                        new {
                            AbillityEntityId = abilityEntityId,
                            AbilityCardId = abilityCardId,
                            AbilitySpeed = abilitySpeed,
                        }),
                    true,
                    node) };
            }
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            return null;
        }
    }
}
